iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/MultiTsFileDeviceIterator.java

```diff
@@ -398,59 +404,43 @@ private void applyModificationForAlignedChunkMetadataList(
-              try {
-                for (IDeviceID device : r.getDevices()) {
-                  // TODO: remove deviceId conversion
-                  long timeLowerBound =
-                      CommonDateTimeUtils.currentTime()
-                          - DataNodeTTLCache.getInstance()
-                              .getTTL(((PlainDeviceID) device).toStringID());
-                  if (r.getStartTime(device) < timeLowerBound) {
-                    list.add(
-                        new Deletion(
-                            CompactionPathUtils.getPath(device)
-                                .concatNode(IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD),
-                            Long.MAX_VALUE,
-                            Long.MIN_VALUE,
-                            timeLowerBound));
-                  }
-                }
-              } catch (IllegalPathException e) {
-                throw new RuntimeException(e);
-              }
-              return list;
-            });
+            r -> new LinkedList<>(ModificationFile.getNormalMods(r).getModifications()));
 
     // construct the input params List<List<Modification>> for QueryUtils.modifyAlignedChunkMetaData
     AlignedChunkMetadata alignedChunkMetadata = alignedChunkMetadataList.get(0);
     List<IChunkMetadata> valueChunkMetadataList = alignedChunkMetadata.getValueChunkMetadataList();
     List<List<Modification>> modificationForCurDevice = new ArrayList<>();
-    List<PartialPath> valueSeriesPaths = new ArrayList<>(valueChunkMetadataList.size());
-    for (int i = 0; i < valueChunkMetadataList.size(); ++i) {
-      modificationForCurDevice.add(new ArrayList<>());
-      IChunkMetadata valueChunkMetadata = valueChunkMetadataList.get(i);
-      valueSeriesPaths.add(
-          valueChunkMetadata == null
-              ? null
-              : CompactionPathUtils.getPath(
-                  currentDevice.left, valueChunkMetadata.getMeasurementUid()));
-    }
-
-    for (Modification modification : modifications) {
-      for (int i = 0; i < valueChunkMetadataList.size(); ++i) {
-        PartialPath path = valueSeriesPaths.get(i);
-        if (path != null && modification.getPath().matchFullPath(path)) {
-          modificationForCurDevice.get(i).add(modification);
+    for (IChunkMetadata valueChunkMetadata : valueChunkMetadataList) {
+      List<Modification> modificationList = new ArrayList<>();
+      modificationForCurDevice.add(modificationList);
+      if (valueChunkMetadata == null) {
+        continue;
+      }
+      PartialPath path =
+          CompactionPathUtils.getPath(
+              currentDevice.getLeft(), valueChunkMetadata.getMeasurementUid());
+      for (Modification modification : modifications) {
+        if (modification.getPath().matchFullPath(path)) {
+          modificationList.add(modification);
         }
       }
+      if (ttlDeletion != null) {
+        modificationList.add(ttlDeletion);
+      }
     }
```

Comment:
emptyList may occupy less memory than empty ArrayList if the list is unmodifiable.
```suggestion
    for (IChunkMetadata valueChunkMetadata : valueChunkMetadataList) {
      if (valueChunkMetadata == null) {
         modificationForCurDevice.add(Collections.emptyList());
        continue;
      }
      List<Modification> modificationList = new ArrayList<>();
      PartialPath path =
          CompactionPathUtils.getPath(
              currentDevice.getLeft(), valueChunkMetadata.getMeasurementUid());
      for (Modification modification : modifications) {
        if (modification.getPath().matchFullPath(path)) {
          modificationList.add(modification);
        }
      }
      if (ttlDeletion != null) {
        modificationList.add(ttlDeletion);
      }
      modificationForCurDevice.add(modificationList.isEmpty() ? Collections.emptyList() : modificationList);
    }
```